tests/integration/tests/charm_tests/openstack_exporter.py

```diff
@@ -256,18 +257,45 @@ def test_openstack_exporter_snap_down(self):
         )
         self.assertEqual(self.leader_unit.workload_status_message, "")
 
+    def test_upstream_exporter_as_resource(self):
+        """Test using the upstream golang as resource will block the unit."""
+        # juju cannot access resources at /tmp, so we create a tmp folder at the current directory
+        with tempfile.TemporaryDirectory(dir="./") as tmpdir:
+            tmp_path = Path(tmpdir)
+            download_cmd = (
+                f"wget -q -P {tmpdir} https://github.com/canonical/openstack-exporter-operator/"
+                "releases/download/rev2/golang-openstack-exporter_amd64.snap"
+            )
+            subprocess.run(download_cmd.split(), check=False)
+
+            resource_file = tmp_path / "golang-openstack-exporter_amd64.snap"
+            model.attach_resource(APP_NAME, "openstack-exporter", str(resource_file))
+            model.block_until_unit_wl_status(
+                self.leader_unit_entity_id, "blocked", timeout=STATUS_TIMEOUT
+            )
+            snaps_installed = [snap["name"] for snap in get_snaps_installed()]
+            self.assertIn(UPSTREAM_SNAP, snaps_installed)
+            expected_msg = (
+                f"{UPSTREAM_SNAP} should not be used anymore. " "Please remove the snap resource."
+            )
+            self.assertEqual(self.leader_unit.workload_status_message, expected_msg)
+
+            # passing an empty file will unblock the unit
+            temp_file_path = Path(tmpdir) / f"{SNAP_NAME}.snap"
+            temp_file_path.touch()
+            model.attach_resource(APP_NAME, "openstack-exporter", str(temp_file_path))
+            model.block_until_unit_wl_status(
+                self.leader_unit_entity_id, "active", timeout=STATUS_TIMEOUT
+            )
+
+
+def get_snaps_installed() -> dict[str, str]:
+    """Get the snaps installed in the leader unit."""
+    cmd = "curl -sS --unix-socket /run/snapd.socket http://localhost/v2/snaps"
+    return json.loads(model.run_on_leader(APP_NAME, cmd).get("Stdout", ""))["result"]
+
 
-def get_snap_exporter_info() -> dict[str, str]:
-    """Get the openstack exporter snap information.
-
-    The snap list expected format as input is:
-    Name                        Version            Rev    Tracking       Publisher   Notes
-    charmed-openstack-exporter  1.7.0-26-g3be9ddb  4      latest/stable  canonical✓  -
-    """
-    results = model.run_on_leader(APP_NAME, f"snap list {SNAP_NAME}").get("Stdout", "").strip()
-    snap_info = results.splitlines()[1].split()
-    return {
-        "version": snap_info[VERSION_COLUMN],
-        "revision": snap_info[REVISION_COLUMN],
-        "channel": snap_info[CHANNEL_COLUMN],
-    }
+def get_snap_info(snap_name: str) -> list[dict[str, str]]:
+    """Get the snap information."""
+    snaps = get_snaps_installed()
+    return [snap for snap in snaps if snap["name"] == snap_name][0]
```

Comment:
I didn't know we can do that 👀. Maybe this is an easier and more efficient way of getting one snap info, e.g.

```
curl -sS --unix-socket /run/snapd.socket http://localhost/v2/find -G -d "name=juju"
```

Reference: https://github.com/canonical/operator-libs-linux/blob/20880b1dfba4f8e38a8026bb06ea9d3d30f6a081/lib/charms/operator_libs_linux/v2/snap.py#L785

Reply:
Nice finding, I didn't know it either.
However I think this is not applicable to this case because I want the information from the snap installed and this gives the information of the snap from the snapstore

Reply:
Thanks, I should double checked 🙂